### PR TITLE
Better environment handling

### PR DESCRIPTION
**Pull request by jonludlam**
Ensures we only ever add items to the environment once.

### CODE REVIEW

File: src/xref2/env.ml [hunk 1]
```diff
@@ -136,6 +166,9 @@ end = struct
 end
 
 type t = {
+  linking : bool;
+  (* True if this is a linking environment - if not,
+     we only put in modules, module types, types, classes and class types *)
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply src/xref2/env.ml:171]: We might be able to split it into two - the 'core' elements and the 'extras' perhaps? I think this is perhaps beyond the scope of this particular PR though.

[the remark this answers is not in the document]
Comment (Julow) [reply src/xref2/env.ml:173]: Alright, let's improve that later.

File: src/xref2/env.ml [hunk 2]
```diff
@@ -112,6 +114,17 @@ end = struct
   let empty = (StringMap.empty, IdMap.empty)
 
   let add ?(shadow = true) kind identifier elem (names, ids) =
+    let check_for_duplicates () =
+      try
+        ignore (IdMap.find (identifier :> Identifier.t) ids);
```
Comment (Julow) [inline src/xref2/env.ml:119]: Could be `if IdMap.mem ...`.

**Review by Julow — APPROVED** [review]
Let's merge :)

File: src/xref2/env.ml (hunk 2 above)
Comment (jonludlam) [reply src/xref2/env.ml:119]: Erm, yes... :-)

Comment (jonludlam) [reply src/xref2/env.ml]: Heh, fixed :-) thanks!

**Pull request merged**